File: python_legacy/tester/scorer.py
# -*- coding: utf-8 -*-
"""模型评分模块：同分类内横向评分 + 排名"""

# 各测试用例权重（满分 90，另 10 分来自速度）
TEST_WEIGHTS = {
    "basic_availability": 30, "chinese_support": 10, "tool_calling": 20,
    "streaming": 10, "think_on": 8, "think_off": 4, "long_context": 8,
    "code_generation": 30, "code_completion": 15, "code_debugging": 30,
    "code_explanation": 15, "basic_embedding": 40, "batch_embedding": 30,
    "query_embedding": 30, "vlm_image_understanding": 30, "multi_image": 15,
    "basic_rerank": 100, "text2image": 100,
    "basic": 20, "math_reasoning": 30, "cot_output": 30, "logic_puzzle": 20,
    "text_via_clip": 50, "image_via_clip": 50, "tts": 100,
}
GRADE_MAP = [(85, "S"), (70, "A"), (55, "B"), (40, "C"), (20, "D"), (0, "F")]


def score_model(results: list) -> dict:
    """计算单个模型的综合评分（0-100）"""
    weighted_sum, total_weight, tps_list, elapsed_list = 0, 0, [], []
    for r in results:
        status = r.get("status", "")
        if status == "skip":
            continue  # skip = 模型不支持该功能，不计入评分分母
        w = TEST_WEIGHTS.get(r.get("test", ""), 0)
        if w:
            total_weight += w
            if status == "pass":
                weighted_sum += w
        tps = r.get("tps", 0)
        if tps and tps > 0:
            tps_list.append(tps)
        elapsed = r.get("elapsed_ms", 0)
        if elapsed and elapsed > 0:
            elapsed_list.append(elapsed)

    base = (weighted_sum / total_weight * 90) if total_weight else 0
    avg_tps = sum(tps_list) / len(tps_list) if tps_list else 0
    # D-2: 对无 TPS 的模型（embedding/reranker），用 elapsed_ms 倒数作速度指标
    avg_elapsed = sum(elapsed_list) / len(elapsed_list) if elapsed_list else 0
    score = round(base, 1)
    grade = next(g for threshold, g in GRADE_MAP if score >= threshold)
    passed = sum(1 for r in results if r.get("status") == "pass")
    skipped = sum(1 for r in results if r.get("status") == "skip")
    return {
        "score": score,
        "grade": grade,
        "avg_tps": round(avg_tps, 1),
        "avg_elapsed": round(avg_elapsed, 1),  # 供 rank_category 使用
        "passed": passed,
        "skipped": skipped,
        "total": len(results),
    }


def rank_category(model_score_map: dict) -> dict:
    """
    输入: {model_id: score_dict}
    输出: {model_id: {rank, score, grade, ...}}，按分类内分数排名
    """
    if not model_score_map:
        return {}

    # 速度归一化加分（类内最快 +10，最慢 +0）
    # TPS 模型（chat/code/vlm）按 avg_tps 归一化
    # D-2: 无 TPS 的模型（embedding/reranker）按 avg_elapsed 倒数归一化
    tps_values = [v["avg_tps"] for v in model_score_map.values() if v["avg_tps"] > 0]
    max_tps = max(tps_values) if tps_values else 0

    if max_tps > 0:
        # 有 TPS 数据的分类（如对话/推理/代码等）。avg_tps == 0 的模型在此分类中不给速度奖励分，防止因流式失败反而拿满分。
        min_elapsed = 0
    else:
        # 纯非 TPS 分类（如 embedding/reranker）。使用 avg_elapsed 倒数归一化速度分。
        elapsed_values = [v["avg_elapsed"] for v in model_score_map.values() if v["avg_elapsed"] > 0]
        min_elapsed = min(elapsed_values) if elapsed_values else 0

    scored = {}
    for mid, s in model_score_map.items():
        if s["avg_tps"] > 0 and max_tps > 0:
            speed_bonus = s["avg_tps"] / max_tps * 10
        elif max_tps == 0 and s["avg_elapsed"] > 0 and min_elapsed > 0:
            # 仅在非 TPS 分类下才使用延迟倒数加分
            speed_bonus = min_elapsed / s["avg_elapsed"] * 10
        else:
            speed_bonus = 0
        final = min(100, round(s["score"] + speed_bonus, 1))
        grade = next(g for threshold, g in GRADE_MAP if final >= threshold)
        scored[mid] = {**s, "score": final, "grade": grade}

    # 排名（分数相同并列）
    sorted_ids = sorted(scored, key=lambda m: scored[m]["score"], reverse=True)
    rank, prev_score, prev_rank = 1, None, 1
    for i, mid in enumerate(sorted_ids):
        if scored[mid]["score"] != prev_score:
            prev_rank = i + 1
        scored[mid]["rank"] = prev_rank
        prev_score = scored[mid]["score"]

    return scored

File: src/tester/scorer.ts
import type { ModelResult, ScoredModel } from "./types.ts";

// 各测试用例权重（满分 90，另 10 分来自速度）
export const TEST_WEIGHTS: Record<string, number> = {
  basic_availability: 30,
  chinese_support: 10,
  tool_calling: 20,
  streaming: 10,
  think_on: 8,
  think_off: 4,
  long_context: 8,
  code_generation: 30,
  code_completion: 15,
  code_debugging: 30,
  code_explanation: 15,
  basic_embedding: 40,
  batch_embedding: 30,
  query_embedding: 30,
  vlm_image_understanding: 30,
  multi_image: 15,
  basic_rerank: 100,
  text2image: 100,
  basic: 20,
  math_reasoning: 30,
  cot_output: 30,
  logic_puzzle: 20,
  text_via_clip: 50,
  image_via_clip: 50,
  tts: 100,
};

export const GRADE_MAP: [number, string][] = [
  [85, "S"],
  [70, "A"],
  [55, "B"],
  [40, "C"],
  [20, "D"],
  [0, "F"],
];

export function getGradeFromScore(score: number): string {
  for (const [threshold, grade] of GRADE_MAP) {
    if (score >= threshold) {
      return grade;
    }
  }
  return "F";
}

/**
 * 计算单个模型的综合评分（0-90 的基础分段）
 */
export function scoreModel(
  results: ModelResult[],
  maxContext?: number,
  paramCount?: number
): ScoredModel {
  let weightedSum = 0;
  let totalWeight = 0;
  const tpsList: number[] = [];
  const elapsedList: number[] = [];

  for (const r of results) {
    const status = r.status || "";
    if (status === "skip") {
      continue; // skip = 模型不支持该功能，不计入评分分母
    }
    const testName = r.test || "";
    const w = TEST_WEIGHTS[testName] || 0;
    if (w > 0) {
      totalWeight += w;
      if (status === "pass") {
        weightedSum += w;
      }
    }
    if (r.tps && r.tps > 0) {
      tpsList.push(r.tps);
    }
    if (r.elapsed_ms && r.elapsed_ms > 0) {
      elapsedList.push(r.elapsed_ms);
    }
  }

  const base = totalWeight > 0 ? (weightedSum / totalWeight) * 90 : 0;
  const avg_tps = tpsList.length > 0 ? tpsList.reduce((a, b) => a + b, 0) / tpsList.length : 0;
  const avg_elapsed = elapsedList.length > 0 ? elapsedList.reduce((a, b) => a + b, 0) / elapsedList.length : 0;

  // 提取 max_context 和 param_count
  let finalMaxContext = maxContext;
  if (finalMaxContext === undefined) {
    for (const r of results) {
      if (typeof r.max_context === "number") {
        finalMaxContext = r.max_context;
        break;
      }
      if (r.meta_json && typeof r.meta_json.max_context === "number") {
        finalMaxContext = r.meta_json.max_context;
        break;
      }
      if (r.meta && typeof r.meta.max_context === "number") {
        finalMaxContext = r.meta.max_context;
        break;
      }
    }
  }

  let finalParamCount = paramCount;
  if (finalParamCount === undefined) {
    for (const r of results) {
      if (typeof r.param_count === "number") {
        finalParamCount = r.param_count;
        break;
      }
      if (r.meta_json && typeof r.meta_json.param_count === "number") {
        finalParamCount = r.meta_json.param_count;
        break;
      }
      if (r.meta && typeof r.meta.param_count === "number") {
        finalParamCount = r.meta.param_count;
        break;
      }
    }
  }

  let score = base;

  // 硬性约束扣分：
  // 1. Max Context Constraint: 若 max_context < 32768，扣除 15 分
  if (finalMaxContext !== undefined && finalMaxContext < 32768) {
    score -= 15;
  }
  // 2. Low Speed Constraint: 若 avg_tps < 3.0，扣除 20 分
  if (avg_tps > 0 && avg_tps < 3.0) {
    score -= 20;
  }

  score = Math.max(0, Math.min(90, Math.round(score * 10) / 10));

  let grade = getGradeFromScore(score);

  // 硬性约束评级上限：
  // 1. Max Context Constraint: 若 max_context < 32768，上限为 B
  if (finalMaxContext !== undefined && finalMaxContext < 32768) {
    if (grade === "S" || grade === "A") {
      grade = "B";
    }
  }
  // 2. Low Speed Constraint: 若 avg_tps < 3.0，上限为 C
  if (avg_tps > 0 && avg_tps < 3.0) {
    if (grade === "S" || grade === "A" || grade === "B") {
      grade = "C";
    }
  }

  const passed = results.filter((r) => r.status === "pass").length;
  const skipped = results.filter((r) => r.status === "skip").length;

  return {
    score,
    grade,
    avg_tps: Math.round(avg_tps * 10) / 10,
    avg_elapsed: Math.round(avg_elapsed * 10) / 10,
    passed,
    skipped,
    total: results.length,
    use_cases: [],
    use_cases_str: "",
    max_context: finalMaxContext,
    param_count: finalParamCount,
  };
}

/**
 * 输入: { model_id: ScoredModel }
 * 输出: { model_id: ScoredModel }，按分类内分数排名且加上速度加分，并应用评级上限
 */
export function rankCategory(modelScoreMap: Record<string, ScoredModel>): Record<string, ScoredModel> {
  const entries = Object.entries(modelScoreMap);
  if (entries.length === 0) {
    return {};
  }

  const scoredModels = entries.map(([_, v]) => v);
  const tpsValues = scoredModels.map((m) => m.avg_tps).filter((v) => v > 0);
  const maxTps = tpsValues.length > 0 ? Math.max(...tpsValues) : 0;

  let minElapsed = 0;
  if (maxTps === 0) {
    const elapsedValues = scoredModels.map((m) => m.avg_elapsed).filter((v) => v > 0);
    minElapsed = elapsedValues.length > 0 ? Math.min(...elapsedValues) : 0;
  }

  const scored: Record<string, ScoredModel> = {};

  for (const [mid, s] of entries) {
    let speedBonus = 0;
    if (maxTps > 0 && s.avg_tps > 0) {
      speedBonus = (s.avg_tps / maxTps) * 10;
    } else if (maxTps === 0 && s.avg_elapsed > 0 && minElapsed > 0) {
      speedBonus = (minElapsed / s.avg_elapsed) * 10;
    }

    let finalScore = s.score + speedBonus;
    finalScore = Math.max(0, Math.min(100, Math.round(finalScore * 10) / 10));

    let grade = getGradeFromScore(finalScore);

    // 重新应用评级上限（因为加了速度分可能让等级回升，所以最终要在此处再做一次上限约束）
    if (s.max_context !== undefined && s.max_context < 32768) {
      if (grade === "S" || grade === "A") {
        grade = "B";
      }
    }
    if (s.avg_tps > 0 && s.avg_tps < 3.0) {
      if (grade === "S" || grade === "A" || grade === "B") {
        grade = "C";
      }
    }

    scored[mid] = {
      ...s,
      score: finalScore,
      grade,
    };
  }

  // 排名计算（同分并列排名）
  const sortedIds = Object.keys(scored).sort((a, b) => scored[b]!.score - scored[a]!.score);
  let prevScore: number | null = null;
  let prevRank = 1;

  for (let i = 0; i < sortedIds.length; i++) {
    const mid = sortedIds[i]!;
    const currentScore = scored[mid]!.score;
    if (currentScore !== prevScore) {
      prevRank = i + 1;
    }
    scored[mid]!.rank = prevRank;
    prevScore = currentScore;
  }

  return scored;
}
